pkg/filesystem/fs/fs.go

```diff
@@ -22,16 +63,732 @@ const (
-	Support(ctx context.Context, labels map[string]string) bool
-	PrepareMetaLayer(ctx context.Context, snapshot storage.Snapshot, labels map[string]string) error
-	PrepareBlobLayer(ctx context.Context, snapshot storage.Snapshot, labels map[string]string) error
-	MountPoint(snapshotID string) (string, error)
-	BootstrapFile(snapshotID string) (string, error)
-	NewDaemonConfig(labels map[string]string) (config.DaemonConfig, error)
-	CleanupBlobLayer(ctx context.Context, key string, async bool) error
+type ImageMode int
+
+const (
+	OnDemand ImageMode = iota
+	PreLoad
+)
+
+func init() {
+	buf := [2]byte{}
+	*(*uint16)(unsafe.Pointer(&buf[0])) = uint16(0xABCD)
+
+	switch buf {
+	case [2]byte{0xCD, 0xAB}:
+		nativeEndian = binary.LittleEndian
+	case [2]byte{0xAB, 0xCD}:
+		nativeEndian = binary.BigEndian
+	default:
+		panic("Could not determine native endianness.")
+	}
+}
+
+type Filesystem struct {
+	meta.FileSystemMeta
+	manager              *process.Manager
+	cacheMgr             *cache.Manager
+	verifier             *signature.Verifier
+	sharedDaemon         *daemon.Daemon
+	daemonCfg            config.DaemonConfig
+	resolver             *Resolver
+	stargzResolver       *stargz.Resolver
+	vpcRegistry          bool
+	daemonBackend        string
+	nydusdBinaryPath     string
+	nydusImageBinaryPath string
+	mode                 Mode
+	logLevel             string
+	logDir               string
+	logToStdout          bool
+	nydusdThreadNum      int
+	imageMode            ImageMode
+	blobMgr              *BlobManager
+}
+
+func getBlobPath(dir string, blobDigest string) (string, error) {
+	digest, err := digest.Parse(blobDigest)
+	if err != nil {
+		return "", errors.Wrapf(err, "invalid layer digest %s", blobDigest)
+	}
+	return filepath.Join(dir, digest.Encoded()), nil
+}
+
+// NewFileSystem initialize Filesystem instance
+func NewFileSystem(ctx context.Context, opt ...NewFSOpt) (*Filesystem, error) {
+	var fs Filesystem
+	for _, o := range opt {
+		err := o(&fs)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if fs.imageMode == PreLoad {
+		fs.blobMgr = NewBlobManager(fs.daemonCfg.Device.Backend.Config.Dir)
+		go func() {
+			err := fs.blobMgr.Run(ctx)
+			if err != nil {
+				log.G(ctx).Warnf("blob manager run failed %s", err)
+			}
+		}()
+	}
+	fs.resolver = NewResolver()
+	// Try to reconnect to running daemons
+	if err := fs.manager.Reconnect(ctx); err != nil {
+		return nil, errors.Wrap(err, "failed to reconnect daemons")
+	}
+	return &fs, nil
+}
+
+func (fs *Filesystem) CleanupBlobLayer(ctx context.Context, key string, async bool) error {
+	if fs.blobMgr == nil {
+		return nil
+	}
+	return fs.blobMgr.Remove(key, async)
+}
+
+func (fs *Filesystem) PrepareBlobLayer(ctx context.Context, snapshot storage.Snapshot, labels map[string]string) error {
+	start := time.Now()
+	defer func() {
+		duration := time.Since(start)
+		log.G(ctx).Infof("total nydus prepare data layer duration %d", duration.Milliseconds())
+	}()
+
+	if fs.imageMode == OnDemand {
```

Comment:
Move it to before the duration calculation? No need to calculate duration for preload mode.

Reply:
Also it's better to check `fs.imageMode == Preload` directly.

Reply:
Fixed